test/chrome/crdpMultiplexing/crdpMultiplexor.test.ts

```diff
@@ -0,0 +1,175 @@
+        webSocketMock.verify(s => s.send(It.isAnyString()), Times.atLeastOnce());
+    });
+
+    test('Multiplexor message handles multiple channels', () => {
+        let callbackPromises: Promise<void>[] = [];
+
+        for (let i = 1; i < 5; i++) {
+            let channel = multiplexor.addChannel("Channel" + i);
+
+            let callbackPromise = new Promise<void>(function (resolve, reject) {
+                let callback = (data: string) => {
+                    Assert.equal(JSON.parse(data).id, i);
+                    resolve();
+                }
+
+                channel.on("message", callback);
+            });
+
+            callbackPromises.push(callbackPromise);
+            multiplexor.send(channel, '{"method":"Console.enable","id":' + i + '}');
+        }
+
+        // wait for all promises
+        return Promise.all(callbackPromises);
+    });
+
+    test('Multiplexor notification delivered to all channels', () => {
+        let callbackPromises: Promise<void>[] = [];
+        let testNotification: string = "Test.notification";
+        let testEnable: string = "Test.enable";
+
+        for (let i = 1; i < 5; i++) {
+            let channel = multiplexor.addChannel("Channel" + i);
+
+            let callbackPromise = new Promise<void>(function (resolve, reject) {
+                let callback = (data: string) => {
+                    let notification = JSON.parse(data);
+                    if (notification.id !== undefined) {
+                        // first callback is the response to test.enable
+                        Assert.equal(notification.id, i);
+                    } else {
+                        // second callback is the Test.notification callback
+                        Assert.equal(notification.method, testNotification);
+                        resolve();
+                    }
+                }
+
+                channel.on("message", callback);
+            });
+
+            // Enable the Test domain so we will receive notifications
+            channel.send('{"method":"' + testEnable + '","id":' + i + '}');
+            callbackPromises.push(callbackPromise);
+        }
+
+        // send a notification to the socket message callbacks
+        Assert.equal(socketMessageCallbacks.length, 1);
+        socketMessageCallbacks[0]('{"method":"' + testNotification + '"}');
+
+        // wait for all promises
+        return Promise.all(callbackPromises);
+    });
+
+    test('Notifications are delayed until domain is enabled', done => {
+        let domain1Enable = "Domain1.enable";
+        let domain2Enable = "Domain2.enable";
+        let domain1Notification = "Domain1.notification";
+        let domain2Notification = "Domain2.notification";
+
+        let domain2Enabled = false;
+
+        let channel = multiplexor.addChannel("channel");
+        channel.on("message", (data: string) => {
+            let notification = JSON.parse(data);
```

Comment:
It's difficult for me to follow this logic... I think it might be easier for me to follow if we logged all the messages in an array, and then we just verified the contents at the end...